receiver/awsxrayreceiver/internal/translator/http.go

```diff
@@ -0,0 +1,64 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package translator
+
+import (
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/translator/conventions"
+	tracetranslator "go.opentelemetry.io/collector/translator/trace"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsxrayreceiver/internal/tracesegment"
+)
+
+func addHTTP(seg *tracesegment.Segment, span *pdata.Span) {
+	span.Status().InitEmpty() // by default this sets the code to `Status_Ok`
+	if seg.HTTP == nil {
+		return
+	}
+
+	// https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/master/exporter/awsxrayexporter/translator/http.go#L47
+	attrs := span.Attributes()
+
+	if req := seg.HTTP.Request; req != nil {
+		// https://docs.aws.amazon.com/xray/latest/devguide/xray-api-segmentdocuments.html#api-segmentdocuments-http
+		addString(req.Method, conventions.AttributeHTTPMethod, &attrs)
+
+		if req.ClientIP != nil {
+			// since the ClientIP is not nil, this means that this segment is generated
+			// by a server serving an incoming request
+			attrs.UpsertString(conventions.AttributeHTTPClientIP, *req.ClientIP)
+			span.SetKind(pdata.SpanKindSERVER)
```

Comment:
HTTP doesn't seem like the right file to be setting the kind. Anyways, isn't it just whether it's a segment or a subsegment and whether it's remote or not?

Reply:
Not sure I understand what you were getting at here.

The main reason I put it here is because the `ClientIP` is the sure way to determine whether this is a server span or client span:
https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#spankind

Reply:
Well we're setting kind in `name.go`. It's too mysterious to also have kind determined in `http.go`, especially since it's not really a feature of HTTP. I think it's ok to peek into the HTTP struct again in a different package and is a win to keep the logic for determining kind together.

For client_ip, I was worried because the data model docs say all fields of HTTP are optional - determining one of the most important fields based on an optional is risky. In practice, I think all our SDKs do set it though so it may be ok. But if we just set kind server for all segments I think it's ok too - while it's possible to have `INTERNAL` segments, none of the other data models support this. So in the xray exporter we just map all server spans to segments

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/master/exporter/awsxrayexporter/translator/segment.go#L201

Reply:
I moved 1 line to `name.go`. For your suggestion on setting default kind to `SpanKindSERVER` for all segments, the main problem is I can not tell when I should set kind to `pdata.SpanKindCLIENT`. Essentially, I don't know what to write for XXX:

```
if XXX is true:
    span.SetKind(pdata.SpanKindCLIENT)
else:
    span.SetKind(pdata.SpanKindSERVER)
```

which is why I wrote it using the optional `ClientIP` field:
```
if seg.HTTP != nil && seg.HTTP.Request != nil && seg.HTTP.Request.ClientIP != nil {
	// since the ClientIP is not nil, this means that this segment is generated
	// by a server serving an incoming request
	span.SetKind(pdata.SpanKindSERVER)
} else {
	span.SetKind(pdata.SpanKindCLIENT)
}
```

Reply:
I have updated the code but can't publish another revision yet since it also contains changes in #702. Still waiting on that PR.